tools/shared/src/main/scala/org/scalajs/core/tools/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -875,17 +879,22 @@ private[optimizer] abstract class OptimizerCore(
           }
         }
 
-      case AsInstanceOf(expr, tpe) =>
+      case AsInstanceOf(expr, typeRef) =>
         pretransformExpr(expr) { texpr =>
-          tpe match {
-            case ClassType(ObjectClass) =>
+          typeRef match {
+            case ClassRef(ObjectClass) =>
               cont(texpr)
             case _ =>
-              // TODO This cast is suspicious
-              if (isSubtype(texpr.tpe.base, tpe.asInstanceOf[Type])) {
+              // TODO The hard conversion of ClassRef into ClassType is suspicious
```

Comment:
Same comment here about subtypeing. Maybe we should lift this into a helper `isSubRefType` or something and document stuff there.